### PR TITLE
Weston(37) Database Setup

### PR DESCRIPTION
**Pull request by WillWeston94**
The first step in this entire process is going to be setting up our framework and making sure we are mapping out the correct function of the tables and how they relate to each other.

Tables show below are stored within the db/data directory 

Merchants 
Items
Customers
Invoices
Transactions
InvoiceItems

Loaded in the above order to maintain the structure of the actual db. Ordering is very important as the references towards the other tables need to be taken into account.

NOTE: The rake tasks and applicable code are stored within the lib/tasks directory 

We are able to rake the CSV files using the foreach and rows functions to pull out each piece of data. 

NOTE: the primary id that has already been created in the CSV has been redacted, so our PostgreSQL automatically sequences the numbers. This can be done either through redacting the row or using a primary key reset statement show below

***** ActiveRecord::Base.connection.reset_pk_sequence! *****

I ran into some issues with duplicates using the reset_pk_sequence so I ended up using the former method of just redacting the row all together so PostgreSQL creates our sequential IDs.

Factory_Bot_Rails: These have been generated and are stored in the spec/factories routing. More info to come. 

ENUM STATUS: This is extremely important to the architecture of our DB. Do NOT under any circumstance alter these in either the invoice model or the invoiceItem model. They are used to change the string literal into an integer so our DB is setup not only correctly but efficiently. Using integers for status codes is standard practice. 

Item Model: I have built in a formatting method for when we pull data from the table. 

Create: Transaction model
Create: merchant_spec model
Create: item_spec model
Create: invoice_spec model
Create: Invoice_item_spec model
Create: customer_spec model
Create: factory transaction
Create: factory merchant
Create: factory item
Create: factory invoice
Create: factory invoiceitem
Create: factory customer
Create: csv_load transactions
Create: csv_load merchants.rake
Create: csv_load items.rake
Create: csv_load invoices.rake
Create: csv_load invoice_items.rake
Create: csv_load customers.rake
Create: csv_load all
Migration: Customers
Migration: InvoiceItems
Migration: Invoices
Migration: Items
Migration: Merchants
Migration: Transactions
Migration: Add Merchant ref to Items
Migration: Add foreign keys to Invoice Items for invoices + items
Migration: Add foreign key to Invoices for customers
Migration: Add foreign key to Transactions for invoices
Migration: Change unit price in items to integer
Migration: Full Schema
Create: Transaction model
Create: Merchant model
Create: Item model and method to define unit price formatting
Create: Invoice model and enum status to swap status strings to integers
Create: InvoiceItem model and enum status to swap status strings to integers
Create: Customer model

### CODE REVIEW

File: app/models/invoice.rb [hunk 1]
```diff
@@ -0,0 +1,3 @@
+class Invoice < ApplicationRecord
+  enum status: { 'in progress' => 0, 'completed' => 1, 'cancelled' => 2 }
```
Comment (coryrpow) [inline app/models/invoice.rb:2]: It looks like these are a more dynamic version of the adopt don't shop status thing. So it's easier to change a status or display only 'in progress' invoices, stuff like that. Cool!

**Review by kicotiel — APPROVED** [review]
We need to get validations and relationships set up and tested in the models, but DB, tasks, etc all look solid!

File: lib/tasks/csv_load.rake [hunk 2]
```diff
@@ -0,0 +1,11 @@
+namespace :csv_load do
```
Comment (coryrpow) [inline lib/tasks/csv_load.rake:1]: So this is how we'll be able to call all six of the csvs? Interested to learn how this works, but that's what I'm seeing when I look at it, hah.

**Review by NSMcEnroe — APPROVED** [review]
Thanks for doing all this setup!  Backing Tyler's comment, we will need to do validations as a group, but otherwise, it looks great.

File: spec/factories/customers.rb [hunk 3]
```diff
@@ -0,0 +1,8 @@
+FactoryBot.define do
+  factory :customer do
```
Comment (coryrpow) [inline spec/factories/customers.rb:2]: Interestingggg, FactoryBot looks cool!

**Review by TheAveryRodriguez — COMMENTED** [review]
You put a lot of effort into the initial setup and organization of the database and associated tasks. The order that you' loaded the tables and the explanation of redacting primary keys for PostgreSQL sequencing shows boss database management. The note about not altering ENUM STATUS is huge. Strong foundation for our projecet!

**Pull request merged**